src/shared/bpf-link.c

```diff
@@ -0,0 +1,20 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+
+#if HAVE_LIBBPF
+#include <bpf/libbpf.h>
+#endif
+
+#include "bpf-link.h"
+
+#if HAVE_LIBBPF
+BPFLink *bpf_link_free(BPFLink *link) {
+        if (link && libbpf_get_error(link) == 0)
+               (void) bpf_link__destroy(link);
```

Comment:
hmm, can you explain? when there's no error you destroy? i dont get it?

Reply:
currently, libbpf will return error code encoded as pointer (kernel convention) if BPF link (and most other "objects") couldn't be constructed. `libbpf_get_error() == 0` is the official way with libbpf to check whether such pointer is a real pointer (it's never NULL). If bpf_link construction failed, there is nothing to destroy. But as mentioned above, starting with libbpf 0.1, bpf_link__destroy() checks that automatically, so if systemd depends on >=0.1, there is no need to do anything special here.